### ISSUE TITLE
Some thoughts about the way forward

### ISSUE THREAD

**Issue by adtac** [adtac/commento#60]
:point_up: [May 3, 2017 8:56 PM](https://gitter.im/commento-dev/commento?at=5909f69c08c00c092add9ee6)

**Comment by GaiusMartius** [comment]
Recent evolution on the questions raised by my original comment:

From @DarkThrone 
>I mean you can fight spam proactively or reactively
>you can identify spam heuristically by it's content
>or you can deter spam sources with anti-spam mechanisms
>the first suffers from false positives, but the later adds friction

My response was as follows:
>I think Sgritta is right in conceptualizing the Spam issue as false positives vs. Friction.
Friction is probably we're we'll come down because the resources to do content based spam filtering are not available to a small open source project. I mean, if we could solve the spam problem without Friction but Disqus can not then they would need much better engineers.

>Realistically the strategy will come down to using something like email but trying to make the user experience as smooth as possible.
@DarkThrone, I would say that if email is not used then you can suggest different Friction mechanisms. It sounds like you know more about it.
To deal with Spam we can have email to verify user plus dynamic rate limiting.
Dynamic rate limiting means that a user can post 3 messages (or whatever number) quickly but subsequent messages will take progressively longer to send.

>Then the question is, what if a user is continually abusive?
I like the idea of IP banning (and other additional means can be suggested too).
One downside of storing IPs is that users may not want the content of their messages stored along with the source of that message.
Because commento will be unique to page, a universally unique ID (UUID) can be produced and appended to a users IP then run through a cryptographic hash. That will make interating through all IPS not possible unless you have the UUID.
An uuid can be produced in golang with https://github.com/satori/go.uuid

>What about stuff like universities where everyone is under the same IP?
In that case the full "IP" can be UUID + email + IP run through a cryptohash. This will be used for all users regardless of where they come from.
For the university example, you would have two separate users with the same IP but different emails. This would result in different hashed IPs so one would be banned and the other would not

Stuart Baker @smbkr asks
>Without email verification though, wouldn't it be trivial for a user to get around the ban by just changing one char in the email they enter?

My response was
>Sure, I think there should be something that is associated with a users identity. It could be an email address or just a site level username that will be verified with an email address
>Also, we don't want to make the user enter the email on every comment so we should set a cookie with some values that identify the user.
>JWTs allow encoding of user information and can be sent with each request. That would avoid the need to maintain session in commento which makes managing interaction with the service much easier. More stateless in other words.
We'll parse the JWT in middleware and it will contain userID info and other pieces of info that are useful (that can be decided as we go forward since it doesn't require them to be determined up front)

**The tl;dr;**
User uses email, or other mechanism, to 'log-in'. This results in getting a JWT sent back that the user puts in a cookie.

Abuse is handled by banning on two separate dimension.
1) User ban: Ban based on email address in JWT token
2) IP ban: Ban is based on Scrambled IP in JWT Token where Scrambled IP = IP + UUID.

----------
There is a question as to whether the UUID should be page level or site level. If you want both than there will need to be separate pieces of data for each which will increase complexity. It is easier to ask whether it makes to ban someone from a page but not the site makes sense.

**Comment by adtac** [comment]
@GaiusMartius Just a heads up.

We've moved all development [to GitLab](https://gitlab.com/commento/commento-ce). I think this issue is outdated, so I'm closing it. If you feel this is still relevant, please feel free to open a new issue in the GitLab repo issue tracker. Cheers! :)
